### PR TITLE
Python: port ReDoS queries from Javascript

### PR DESCRIPTION
**Pull request by yoff**
This PR ports the very succesful ReDoS queries from Javascript. In order to do so it first:
 - exposes a few predicates in our regex parser (and fixes a few issues)
 - builds a parse-tree-view on top
 - refactors the js queries to admit identical files

Some tests are included here, more can be found [in the scratch branch](https://github.com/github/codeql/pull/6038).

This work was greatly helped by @erik-krogh who contributed numerous fixes along the way as well as making parsed regexes viewable in the AST-viewer and making the `toUnicode` predicate available on `String`s. And he enthusiastically triaged several results from the run-on-alls.

Some things that might still need to be tweaked:
 - Python (and Ruby) have extra anchors that should probably be accounted for in the queries, specifically during prefix generation.
 - The value of constants which are anchors should also be considered so they do not get confused with character classes.
 - It would be nice to add a test file including all the currently triaged results.

### CODE REVIEW

File: python/ql/src/semmle/python/RegexTreeView.qll [hunk 1]
```diff
@@ -445,7 +445,7 @@ class RegExpEscape extends RegExpNormalChar {
     or
     this.getUnescaped() = "t" and result = "\t"
     or
-    this.getUnescaped() = "f" and result = "\f"
```
Comment (yoff) [inline python/ql/src/semmle/python/RegexTreeView.qll]: I wonder if this is sustainable, or if we should elide this case until more escapes are supported by CodeQL...

Comment (yoff) [reply python/ql/src/semmle/python/RegexTreeView.qll]: Elided for now..

Comment (erik-krogh) [reply python/ql/src/semmle/python/RegexTreeView.qll]: Use int.toUnicode()?

**Comment by yoff** [comment]
DCA run [here](https://github.com/dsp-testing/yoff-dca/issues/21)

File: javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+ *   * The over-approximate step relation allows transitions along pairs of
+ *     abstract input symbols where the symbols have overlap in the characters they accept.
+ *   * Once a trace of pairs of abstract input symbols that leads from a fork
+ *     back to itself has been identified, we attempt to construct a concrete
+ *     string corresponding to it, which may fail.
+ *   * Lastly we ensure that any state reached by repeating `n` copies of `w` has
+ *     a suffix `x` (possible empty) that is most likely __not__ accepted.
+ */
+
+import ReDoSUtil
```
Comment (aschackmull) [inline javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll:65]: Driveby comment: Make as much as possible `private`. The import plus most of the classes and predicates all ought to be private.

File: python/ql/src/semmle/python/regex.qll [hunk 3]
```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
...
+
+  /** result is true for those start chars that actually mark a start of a char set. */
+  boolean char_set_start(int pos) {
+    exists(int index |
+      char_set_delimiter(index, pos) = true and
+      (
+        index = 1 and result = true // if a '[' is first in the string (among brackets), it starts a char set
+        or
+        index > 1 and
+        not char_set_delimiter(index - 1, _) = false and
```
Comment (tausbn) [inline python/ql/src/semmle/python/regex.qll:138]: I'm a bit curious about this line. Why not just `char_set_delimiter(index - 1, _) = true`? As far as I can tell, `char_set_delimiter` cannot fail for `index - 1` if it has already succeeded for `index` (though I may have missed something).

Also, it would be great if each disjunct had a brief comment about which case it is handling.

File: python/change-notes/2021-07-28-port-RoDoS-queries.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* Ported _Inefficient regular expression_ (`py/redos`) query from javascript.
+* Ported _Polynomial regular expression used on uncontrolled data_ [`py/polynomial-redos`] query from javascript.
```
Comment (RasmusWL) [inline python/change-notes/2021-07-28-port-RoDoS-queries.md:3]: Not sure customers will care that we ported this, but I think it's fair to highlight that this is already available in JS. (but this is mostly just a nitpick)

```suggestion
* Added _Inefficient regular expression_ (`py/redos`) query, which is already available in JavaScript.
* Added _Polynomial regular expression used on uncontrolled data_ (`py/polynomial-redos`), which is already available in JavaScript.
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
I looked through the first commit and tried to understand the improvements you made to our regex library... I was taken by surprise by how complicated was for me to understand. Must have been a huge effort to actually make these things :muscle: 

I made a few very detailed comments on that first commit, and have glanced through the other commits and made a few comments along the way.

I think it would have been ideal to have such detailed explanatory comments on the whole `regex.qll`, but I think we need to evaluate whether that's actually going to be worth the effort (even just for the new parts added). I suggest we talk this over "in-person", once you've had a chance to look through my comments.

I think a few of the non-private member-predicates of the `RegexString` class have been changed (like `qualifiedItem`/`qualifiedPart` in https://github.com/github/codeql/pull/6175/commits/21007d21f4b53080db0d2dfa44be98bfcd86cb46), so technically that means if someone is using these, their code will break if we push this update :disappointed: So I guess we should be following our normal deprecation policy on this as well, although it is a bit burdensome. We should consider whether new versions should be marked with `INTERNAL: Do not use.`, which can allow us some more freedom (if for example, the predicates are only exposed to be able to test things).

File: python/ql/src/semmle/python/regex.qll [hunk 5]
```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
 abstract class RegexString extends Expr {
   RegexString() { (this instanceof Bytes or this instanceof Unicode) }
 
+  override string toString() {
+    result = this.(Bytes).getText()
+    or
+    result = this.(Unicode).getText()
+  }
+
+  /** result is true for those start chars that actually mark a start of a char set. */
```
[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/semmle/python/regex.qll:130]: Another problem is on line 154, where we recursively invoke `result = char_set_start(p2).booleanNot()`. Rewriting this to not use booleans would mean rewriting this line as `not char_set_start(p2)`, but then we have introduced negative recursion.

**Review by tausbn — CHANGES_REQUESTED** [review]
Still chewing through this PR, but I thought I would give some interim comments. (I am happy to see that RasmusWL has addressed some of them already. :slightly_smiling_face:)

File: javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll (hunk 2 above)
Comment (yoff) [reply javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll:65]: Good point, thanks.

**Comment by yoff** [comment]
DCA run without attempting to use special CLI [here](https://github.com/dsp-testing/yoff-dca/tree/run/python-port-ReDoS-smoke-test-smoke-test-1/reports).

File: python/ql/src/semmle/python/regex.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/regex.qll:130]: Clearly I forgot to write proper comments here. This is following the same pattern as `predicate escapingChar(int pos) { this.escaping(pos) = true }` and `private boolean escaping(int pos)` to avoid negative recursion. The pattern is used again in scanning through char sets to find character ranges. Both of these instances are a bit more elaborate than `escaping` and therefor the pattern is likely harder to detect without documentation.

File: python/ql/src/semmle/python/regex.qll [hunk 6]
```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
...
+            )
+          then
+            exists(int p2 | char_set_delimiter(index - 2, p2) = true |
+              result = char_set_start(p2).booleanNot() // ...if it is not the first in a char set
+            )
+          else result = true
+        )
+      )
+    )
+  }
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/regex.qll:160]: While these comments are correct, and an improvement, and I appreciate that you have done much work for me here, I am slightly unhappy with the wording "special handling of cases such as".

I see the logic as more universal: A charset runs from a starting bracket to an ending bracket. Then can come a number of ending brackets, but the next starting bracket will start a charset. So we "just" have to check that it is preceded by an ending bracket (and one that does not happen to be invalid, here is the special casing).

I am still tempted to accept your suggestion, though :-)

File: python/ql/test/library-tests/regex/charRangeTest.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import re
+
+re.compile(r'[]-[]') #$ charRange=1:2-3:4
+re.compile(r'[---]') #$ charRange=1:2-3:4
+re.compile(r'[\---]') #$ charRange=1:3-4:5
+re.compile(r'[--\-]') #$ charRange=1:2-3:5
+re.compile(r'[\--\-]') #$ charRange=1:3-4:6
+re.compile(r'[0-9-A-Z]') #$ charRange=1:2-3:4 charRange=5:6-7:8
+re.compile(r'[0\-9-A-Z]') #$ charRange=4:5-6:7
+re.compile(r'[0--9-A-Z]') #$ charRange=1:2-3:4 charRange=4:5-6:7
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/test/library-tests/regex/charRangeTest.py:10]: True, invalid ranges do not compile. I guess these cannot occur as typos, then, and perhaps the examples are not as illustrative as I thought...

[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/test/library-tests/regex/charRangeTest.py:10]: We could perhaps add this logic to `failedToParse`.

File: python/ql/src/Security/CWE-730/PolynomialBackTracking.ql [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import python
+import semmle.python.regex.SuperlinearBackTracking
+
+from PolynomialBackTrackingTerm t
+where t.getLocation().getFile().getBaseName() = "KnownCVEs.py"
+select t.getRegex(), t, t.getReason()
```
Comment (RasmusWL) [inline python/ql/src/Security/CWE-730/PolynomialBackTracking.ql:6]: This one looks more like a file that should be under `test/` and not `src/`?

File: python/ql/src/semmle/python/regex.qll [hunk 9]
```diff
@@ -143,6 +143,26 @@ abstract class RegexString extends Expr {
...
   }
 
+  /**
+   * Holds if the character set starting at `charset_start` contains a character range
+   * with lower bound found between `start` and `lower_end`
+   * and upper bound found between `upper_start` and `end`.
+   */
+  predicate charRange(int charset_start, int start, int lower_end, int upper_start, int end) {
+    // mirror logic from `simpleCharacter`
+    exists(int x, int y |
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/regex.qll]: This particular implementation is only for illustrative purposes (there was no such predicate to test). A better implementation is introduced later.

File: python/ql/src/semmle/python/regex.qll [hunk 10]
```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
...
+            )
+          else result = true
+        )
+      )
+    )
+  }
+
+  /** result denotes if the index is a left bracket */
+  boolean char_set_delimiter(int index, int pos) {
+    pos = rank[index](int p | this.nonEscapedCharAt(p) = "[" or this.nonEscapedCharAt(p) = "]") and
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/regex.qll:185]: Intended in the sense that I did not see a reason to change the default indexing...

**Comment by yoff** [comment]
DCA run actually running ReDoS [here](https://github.com/dsp-testing/yoff-dca/issues/23). Performance looks fine. I do not know if the new results are due to duplication, they look fine.

File: python/ql/src/Security/CWE-730/PolynomialBackTracking.ql (hunk 8 above)
Comment (yoff) [reply python/ql/src/Security/CWE-730/PolynomialBackTracking.ql:6]: It does, yes..

File: config/identical-files.json [hunk 11]
```diff
@@ -448,5 +448,17 @@
...
+    "javascript/ql/src/semmle/javascript/security/performance/ReDoSUtil.qll",
+    "python/ql/src/semmle/python/regex/ReDoSUtil.qll"
+  ],
+  "ReDoS Exponential Python/JS": [
+    "javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll",
+    "python/ql/src/semmle/python/regex/ExponentialBackTracking.qll"
+  ],
+  "ReDoS Polynomial Python/JS": [
+    "javascript/ql/src/semmle/javascript/security/performance/SuperlinearBackTracking.qll",
+    "python/ql/src/semmle/python/regex/SuperlinearBackTracking.qll"
```
[the remark this answers is not in the document]
Comment (yoff) [reply config/identical-files.json]: Done

File: python/ql/src/semmle/python/regex.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/semmle/python/regex.qll]: I have added more elaborate comments to this effect now.

**Review by tausbn — APPROVED** [review]

**Review by RasmusWL — APPROVED** [review]

**Review by asgerf — APPROVED** [review]
JS changes lgtm

**Comment by tausbn** [comment]
:ship: :rocket: :boom:

File: javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll (hunk 2 above)
Comment (aschackmull) [reply javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll:65]: So this wasn't actually done?

Comment (yoff) [reply javascript/ql/src/semmle/javascript/security/performance/ExponentialBackTracking.qll:65]: I thought I had done this, but you are right; PR [here](https://github.com/github/codeql/pull/6200).

**Review by duongxuantuan154 — COMMENTED** [review]
- [x] @

**Review by duongxuantuan154 — COMMENTED** [review]
- [ ] @
